internal/resources/gitrepository/data_source_git_repository.go

```diff
@@ -5,10 +5,198 @@ SPDX-License-Identifier: MPL-2.0
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/gitrepository/spec"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/gitrepository/status"
+)
 
 func DataSourceGitRepository() *schema.Resource {
 	return &schema.Resource{
-		Schema: gitRepositorySchema,
+		Schema:      gitRepositorySchema,
+		ReadContext: dataSourceGitRepositoryRead,
+	}
+}
+
+func dataSourceGitRepositoryRead(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
+	config := m.(authctx.TanzuContext)
+
+	gitRepositoryName, ok := d.Get(nameKey).(string)
+	if !ok {
+		return diag.Errorf("unable to read git repository name")
+	}
+
+	gitRepositoryNamespaceName, ok := d.Get(namespaceNameKey).(string)
+	if !ok {
+		return diag.Errorf("unable to read git repository namespace name")
+	}
+
+	gitRepositoryOrgID, _ := d.Get(orgIDKey).(string)
+	scopedFullnameData := scope.ConstructScope(d, gitRepositoryName, gitRepositoryNamespaceName, gitRepositoryOrgID)
+
+	if scopedFullnameData == nil {
+		return diag.Errorf("Unable to get Tanzu Mission Control git repository entry; Scope full name is empty")
+	}
+
+	err := enableContinuousDelivery(&config, scopedFullnameData, common.ConstructMeta(d))
+	if err != nil {
+		return diag.FromErr(errors.Wrapf(err, "Unable to get Tanzu Mission Control git repository entry, name : %s", gitRepositoryName))
+	}
+
+	UID, meta, atomicSpec, clusterScopeStatus, clusterGroupScopeStatus, err := retrieveGitRepositoryUIDMetaAndSpecFromServer(config, scopedFullnameData, d, gitRepositoryName)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// always run
+	d.SetId(UID)
+
+	if err := d.Set(common.MetaKey, common.FlattenMeta(meta)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var (
+		flattenedSpec   []interface{}
+		flattenedStatus interface{}
+	)
+
+	switch scopedFullnameData.Scope {
+	case commonscope.ClusterScope:
+		flattenedSpec = spec.FlattenSpecForClusterScope(atomicSpec)
+		flattenedStatus = status.FlattenStatusForClusterScope(clusterScopeStatus)
+	case commonscope.ClusterGroupScope:
+		clusterGroupScopeSpec := &gitrepositoryclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdGitrepositorySpec{
+			AtomicSpec: atomicSpec,
+		}
+		flattenedSpec = spec.FlattenSpecForClusterGroupScope(clusterGroupScopeSpec)
+		flattenedStatus = status.FlattenStatusForClusterGroupScope(clusterGroupScopeStatus)
+	}
+
+	if err := d.Set(spec.SpecKey, flattenedSpec); err != nil {
+		return diag.FromErr(err)
 	}
+
+	if err := d.Set(statusKey, flattenedStatus); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+// nolint: gocognit
+func retrieveGitRepositoryUIDMetaAndSpecFromServer(config authctx.TanzuContext, scopedFullnameData *scope.ScopedFullname, d *schema.ResourceData, gitRepositoryName string) (
+	string,
+	*objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta,
```

Comment:
too many parameters in the return.. can you make this better

Reply:
Updated.